src/plugins/dashboard/public/application/utils/create_dashboard_container.tsx

```diff
@@ -0,0 +1,522 @@
+  return isEmptyState ? (
+    <DashboardEmptyScreen
+      {...getEmptyScreenProps(
+        shouldShowEditHelp,
+        isEmptyInReadOnlyMode,
+        appStateContainer,
+        container,
+        services
+      )}
+    />
+  ) : null;
+};
+
+const setCurrentIndexPatterns = (dashboardContainer: DashboardContainer) => {
+  let panelIndexPatterns: IndexPattern[] = [];
+  dashboardContainer.getChildIds().forEach((id) => {
+    const embeddableInstance = dashboardContainer.getChild(id);
+    if (isErrorEmbeddable(embeddableInstance)) return;
+    const embeddableIndexPatterns = (embeddableInstance.getOutput() as any).indexPatterns;
+    if (!embeddableIndexPatterns) return;
+    panelIndexPatterns.push(...embeddableIndexPatterns);
+  });
+  panelIndexPatterns = uniqBy(panelIndexPatterns, 'id');
+  return panelIndexPatterns;
+};
+
+const getDashboardInputFromAppState = (
+  appStateData: DashboardAppState,
+  services: DashboardServices,
+  savedDashboardId?: string
+) => {
+  const { data, dashboardConfig } = services;
+  const embeddablesMap: {
+    [key: string]: DashboardPanelState;
+  } = {};
+  appStateData.panels.forEach((panel: SavedDashboardPanel) => {
+    embeddablesMap[panel.panelIndex] = convertSavedDashboardPanelToPanelState(panel);
+  });
+
+  const lastReloadRequestTime = 0;
+  return {
+    id: savedDashboardId || '',
+    filters: data.query.filterManager.getFilters(),
+    hidePanelTitles: appStateData.options.hidePanelTitles,
+    query: data.query.queryString.getQuery(),
+    timeRange: data.query.timefilter.timefilter.getTime(),
+    refreshConfig: data.query.timefilter.timefilter.getRefreshInterval(),
+    viewMode: appStateData.viewMode,
+    panels: embeddablesMap,
+    isFullScreenMode: appStateData.fullScreenMode,
+    isEmptyState:
+      getShouldShowEditHelp(appStateData, dashboardConfig) ||
+      getShouldShowViewHelp(appStateData, dashboardConfig) ||
+      shouldShowUnauthorizedEmptyState(appStateData, services),
+    useMargins: appStateData.options.useMargins,
+    lastReloadRequestTime,
+    title: appStateData.title,
+    description: appStateData.description,
+    expandedPanelId: appStateData.expandedPanelId,
+    timeRestore: appStateData.timeRestore,
+  };
+};
+
+const getChangesForContainerStateFromAppState = (
+  currentContainer: DashboardContainer,
+  appStateDashboardInput: DashboardContainerInput
+) => {
+  if (!currentContainer || isErrorEmbeddable(currentContainer)) {
+    return appStateDashboardInput;
+  }
+
+  const containerInput = currentContainer.getInput();
+  const differences: Partial<DashboardContainerInput> = {};
+
+  // Filters shouldn't  be compared using regular isEqual
+  if (
+    !opensearchFilters.compareFilters(
+      containerInput.filters,
+      appStateDashboardInput.filters,
+      opensearchFilters.COMPARE_ALL_OPTIONS
+    )
+  ) {
+    differences.filters = appStateDashboardInput.filters;
```

Comment:
Would it be safer to have a shallow copy or do we need them to be point to the same reference?

```suggestion
    differences.filters = [...appStateDashboardInput.filters];
```

Reply:
Copied from above

I have created a comment here: https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3365#issuecomment-1627593824.

This logic was ported over, the dashboards plugin is a pretty complex plugin since it's a container embeddable that is managing its state and its children states so previous attempts to clean up certain logic has caused edge case failures.